core/src/main/java/org/kiwix/kiwixmobile/core/main/AddNoteDialog.kt

```diff
@@ -0,0 +1,475 @@
+      articleNoteFileName = getArticleNotefileName()
+      zimNotesDirectory = "$NOTES_DIRECTORY$zimNoteDirectoryName/"
+    } else {
+      onFailureToCreateAddNoteDialog()
+    }
+  }
+
+  private fun onFailureToCreateAddNoteDialog() {
+    context.toast(R.string.error_file_not_found, Toast.LENGTH_LONG)
+    closeKeyboard()
+    requireFragmentManager().beginTransaction().remove(this).commit()
+  }
+
+  override fun onCreateView(
+    inflater: LayoutInflater,
+    container: ViewGroup?,
+    savedInstanceState: Bundle?
+  ): View {
+    super.onCreateView(inflater, container, savedInstanceState)
+    root = inflater.inflate(R.layout.dialog_add_note, container, false)
+    return root
+  }
+
+  private val zimNoteDirectoryName: String
+    get() {
+      val noteDirectoryName = getTextAfterLastSlashWithoutExtension(zimFileName!!)
+      return (if (noteDirectoryName.isNotEmpty()) noteDirectoryName else zimFileTitle)!!
+    }
+
+  private fun getArticleNotefileName(): String {
+    // Returns url of the form: "content://org.kiwix.kiwixmobile.zim.base/A/Main_Page.html"
+    val articleUrl = (activity as WebViewProvider?)?.getCurrentWebView()?.url
+    var noteFileName = ""
+    if (articleUrl == null) {
+      onFailureToCreateAddNoteDialog()
+    } else {
+      noteFileName = getTextAfterLastSlashWithoutExtension(articleUrl)
+    }
+    return (if (noteFileName.isNotEmpty()) noteFileName else articleTitle)!!
+  }
+
+  private fun getTextAfterLastSlashWithoutExtension(path: String): String {
+    /* That's about exactly what it does.
+     *
+     * From ".../Kiwix/granbluefantasy_en_all_all_nopic_2018-10.zim", returns "granbluefantasy_en_all_all_nopic_2018-10"
+     * From "content://org.kiwix.kiwixmobile.zim.base/A/Main_Page.html", returns "Main_Page"
+     * For null input or on being unable to find required text, returns null
+     * */
+    val rightmostSlash = path.lastIndexOf('/')
+    val rightmostDot = path.lastIndexOf('.')
+    return if (rightmostSlash > -1 && rightmostDot > -1) {
+      path.substring(
+        rightmostSlash + 1, if (rightmostDot > rightmostSlash) rightmostDot else path.length
+      )
+    } else ""
+    // If couldn't find a dot and/or slash in the url
+  }
+
+  // Override onBackPressed() to respond to user pressing 'Back' button on navigation bar
+  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+    return object : Dialog(requireContext(), theme) {
+      override fun onBackPressed() {
+        exitAddNoteDialog()
+      }
+    }
+  }
+
+  private fun exitAddNoteDialog() {
+    if (noteEdited) {
+      alertDialogShower?.show(
+        KiwixDialog.NotesDiscardConfirmation,
+        ::dismissAddNoteDialog
+      )
+    } else {
+      // Closing unedited note dialog straightaway
+      dismissAddNoteDialog()
+    }
+  }
+
+  private fun disableMenuItems() {
+    if (toolbar.menu != null) {
+      val saveItem = toolbar.menu.findItem(R.id.save_note)
```

Comment:
we repeat the finding of these items a great deal. Have class level properties for them `lateinit` or nullable, whichever works in this scenario and assign them after the menu gets inflated